### DISCUSSION TITLE
怎样导出自定义组件的函数或变量

### DISCUSSION THREAD

**Discussion by unmagic** [vue-mini/vue-mini#122] [Q&A]
小程序的自定义组件提供了, 和来导出组件的函数或变量，请问使用vue-mini怎么实现导出

小程序的导出
`behaviors: ['wx://component-export'],
export(){} `

**Comment by yangmingshan** [comment]
你的描述过于简短，无法理解你的需求。但是 Vue Mini 是原生小程序的超集，你仍然可以使用 behaviors

**Reply by unmagic** [reply]
<img width="962" alt="image" src="https://github.com/user-attachments/assets/b2b367fa-3419-4dad-b1d6-eb16abe44148">

小程序自定义组件原生的导出方式是上图所示，问下，如果我想导出的是一个在setup中定义的响应式变量，在vue-mini中实现的自定义组件怎么导出该变量

**Reply by yangmingshan** [reply]
这是个反模式，更好的做法是把 data 提升到父组件中，向下传递给子组件，而不是反过来。

**Reply by unmagic** [reply]
那再问下能在自定义组件的setup中获取该自定义组件的data中的数据吗，怎么获取，我看setup中this不可用啊

**Reply by yangmingshan** [reply]
setup 的第二个参数 context 里面有 selectComponent，文档有写。

**Reply by unmagic** [reply]
是在A组件中取他自己的data，不是取子组件的

**Reply by yangmingshan** [reply]
你是在用混合模式吗？这种情况建议把 data 挪到 setup 里面来。

**Reply by unmagic** [reply]
我把data挪到setup里面去的话，我就没法导出这个data里的数据了啊

**Reply by yangmingshan** [reply]
我还是不明白你为什么需要反向传递数据，你的原始需求是什么？提升数据不行吗？

**Reply by yangmingshan** [reply]
setup 的响应式数据最终也在 data 上面，所以你真的需要也可以将它们导出。

**Reply by unmagic** [reply]
> 这是个反模式，更好的做法是把 data 提升到父组件中，向下传递给子组件，而不是反过来。

我可以使用这种方式来解决当前问题，但小程序自定义组件本身是支持导出内部函数或变量的，用了vue-mini后setup与自己的data，两者间无法相互通信，导致我如果想导出自定义组件的内部函数，只能在小程序的原生写法和vue-mini写法中二选一。

**Reply by unmagic** [reply]
我觉得应该在setup的context中再增加一个导出函数，该导出函数可导出在setup中声明的变量或函数，可以完美的解决我现在遇到的问题

**Reply by yangmingshan** [reply]
> setup 的响应式数据最终也在 data 上面，所以你真的需要也可以将它们导出。

你真的有需要也是可以的。但我还是认为这是反模式，建议尽量避免使用，除非没办法。

**Reply by yangmingshan** [reply]
并不需要导出函数，`behaviors: ['wx://component-export']` 是用于自定义导出内容，不用它默认就是全量导出。

**Reply by yangmingshan** [reply]
你可能还是没明白：如果你真的需要，你**可以**导出 setup 里面声明的 data
